packages/flutter_test/lib/src/controller.dart

```diff
@@ -31,12 +225,33 @@ const String _defaultPlatform = kIsWeb ? 'web' : 'android';
 ///
 /// Concrete subclasses must implement the [pump] method.
 abstract class WidgetController {
-  /// Creates a widget controller that uses the given binding.
-  WidgetController(this.binding);
+  /// Creates a widget controller that uses the given binding. A [SemanticsController]
+  /// will be created for [WidgetController.semantics] by default, but a custom
+  /// implementation can be provided for advanced use cases.
+  WidgetController(this.binding)
+    : _semantics = SemanticsController._(binding);
 
   /// A reference to the current instance of the binding.
   final WidgetsBinding binding;
 
+  /// Provides access to a [SemanticsController] for testing anything related to
+  /// the [Semantics] tree. Accessibility services, search engines, and other
+  /// analysis tools all make use of the [Semantics] tree to determine the
+  /// meaning of an application.
+  ///
+  /// If semantics has been disabled for the test, this will throw a [StateError].
```

Comment:
maybe mention one can use WidgetTester.ensureSemantics() to enable semantics

Reply:
I'd actually like to avoid mentioning that since it's not necessary by default. `testWidgets` has semantics enabled by default, so a user would have to specifically _disable_ semantics before they'd run into this issue.

That said, the `StateError` that's thrown if semantics is disabled does mention how to enable semantics through both the `testWidgets` method and the `WidgetController.ensureSemantics()` methods.

Is that sufficient?

Reply:
ah interesting, I don't know it was on by default. we should probably remove the `tester.ensureSemantics()` in our tests as well at some point. Those don't seem to do anything.

Reply:
Yeah, I wasn't aware either actually. goderbaur pointed it out in [this convo](https://github.com/flutter/flutter/pull/113133#discussion_r991410710), but I had always assumed it was needed before that. I was planning on removing unneeded calls while working on the deprecation in #112670 since they're somewhat related.